lib/core/include/irods/configuration_hook_manager.hpp

```diff
@@ -0,0 +1,43 @@
+#ifndef IRODS_CONFIGURATION_HOOK_MANAGER_HPP
+#define IRODS_CONFIGURATION_HOOK_MANAGER_HPP
+
+/// \file
+
+#include <nlohmann/json.hpp>
+
+#include <vector>
+#include <string>
+
+namespace irods::experimental::configuration_reload
+{
+    /// \brief Hooks are functions called with the pathname of the json value that has changed, as well as the new
+    /// operation performed, removal versus replacement versus insertion etc
+    ///
+    /// Hooks are invoked in two stages for each entry in a list of json patches
+    ///
+    /// The first is the "*CATCHALL*" stage, where hooks that are run for every
+    /// configuration change occurs. The second is based on the specific json
+    /// path of the changed configuration value.
+    ///
+    /// For example, if you change the delay server's sleep duration, the
+    /// "*CATCHALL*" hooks will be run first in order of addition, followed by
+    /// the hooks associated with
+    /// "/advanced_settings/delay_server_sleep_time_in_seconds", also in order
+    /// of addition.
+    using hook = std::function<void(const std::string&, const nlohmann::json&)>;
+
+    /// \brief Adds a hook to be run whenever a key on the list of hooks changes.
+    ///        \see hook
+    /// \param [in] hook_triggers An array or container of json paths to match.
+    ///
+    ///             The special value "*CATCHALL*" matches any property that has changed and gets run first
+    /// \param [in] hook The function to run when the json path matches
+    void add_hook(const auto& hook_triggers, hook&& hook);
```

Comment:
Should we add a unit test or three for this hook system? I don't see where it is being used (maybe I missed it), so we should at least test it if we are going to include it. It would be fine if we just create an issue for this so we can move along since this is a main-only change.

Reply:
Yes, we should add tests for this. However, due to time, let's capture that in a new issue.

Reply:
Honestly it wasn't too hard to write the test for this